bftengine/src/bftengine/KeyManager.h

```diff
@@ -1,47 +1,50 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
 #pragma once
 
-#include "Serializable.h"
 #include "InternalBFTClient.h"
+#include "KeyStore.h"
 
 class KeyManager {
  public:
-  struct KeyExchangeMsg : public concord::serialize::SerializableFactory<KeyExchangeMsg> {
-    std::string key;
-    std::string signature;
-    int repID;
-    KeyExchangeMsg(){};
-    KeyExchangeMsg(std::string key, std::string signature, int repID);
-    std::string toString() const;
-    static KeyExchangeMsg deserializeMsg(const char* serStr, const int& size);
-
-   protected:
-    const std::string getVersion() const;
-    void serializeDataMembers(std::ostream& outStream) const;
-    void deserializeDataMembers(std::istream& inStream);
-  };
+  static KeyManager& get(InternalBFTClient* cl = nullptr, const int id = 0, const uint32_t clusterSize = 0) {
+    static KeyManager km{cl, id, clusterSize};
+    return km;
+  }
+
+  void sendKeyExchange();
+  std::string onKeyExchange(KeyExchangeMsg& kemsg, const uint64_t& sn);
+  void onCheckpoint(const int& num);
+  void registerForNotification(IKeyExchanger* ke);
+  KeyExchangeMsg replicaKey(const uint16_t& repID) const;
+
+  std::atomic_bool keysExchanged{false};
 
  private:
-  int repID_{};
+  KeyManager(InternalBFTClient* cl, const int& id, const uint32_t& clusterSize);
+
+  uint16_t repID_{};
   uint32_t clusterSize_{};
   std::set<NodeIdType> exchangedReplicas_;  // TODO to/from Storage
   std::string generateCid();
   // Raw pointer is ok, since this class does not manage this resource.
   InternalBFTClient* client_{nullptr};
 
-  KeyManager(InternalBFTClient* cl, const int& id, const uint32_t& clusterSize);
+  std::vector<IKeyExchanger*> registryToExchange_;
```

Comment:
`KeyManager` does not own `IKeyExchanger*`, is it right?

Reply:
right